### PR TITLE
Add unstack flag to get_array method

### PR DESCRIPTION
**Pull request by daviddoji**
Hopefully this will solve the reindexing ValueErrors when the pulseIds have all the same values for a particular run

### CODE REVIEW

File: extra_data/components.py [hunk 1]
```diff
@@ -328,6 +331,8 @@ def get_array(self, key, pulses=by_index[:]):
           Select the pulses to include from each train. by_id selects by pulse
           ID, by_index by index within the data being read. The default includes
           all pulses. Only used for per-train data.
+        unstack: bool
+          Wheter or not to separate train and pulse dimensions
```
Comment (takluyver) [inline extra_data/components.py:335]:
```suggestion
          Whether to separate train and pulse dimensions.
```

**Comment by takluyver** [comment]
I'm still trying to think if there's a better name for the public parameter. `unstack` is the Xarray method we call to achieve it, but I'm not sure the meaning is obvious from the outside.

**Comment by daviddoji** [comment]
> I'm still trying to think if there's a better name for the public parameter. `unstack` is the Xarray method we call to achieve it, but I'm not sure the meaning is obvious from the outside.

Either that or "separate_dims", which is basically what it basically does.

**Comment by takluyver** [comment]
A bit vague, though. `separate_pulse_dim` or `pulse_dim` (it creates a separate pulse dimension)? `align_trains` or `align_pulses` (it lines up the same pulse across different trains)? :-S

**Comment by daviddoji** [comment]
I would prefer "align_pulses"

**Comment by takluyver** [comment]
OK, I'm happy enough with that. We should add a test of passing `False` - see the other tests in `test_components.py`.

File: extra_data/components.py [hunk 2]
```diff
@@ -608,7 +614,7 @@ def _get_pulse_data(self, source, key, tid):
             # h5py fancy indexing needs a list, not an ndarray
             data_positions = list(first + positions)
 
-        return _guess_axes(ds[data_positions], train_pulse_ids)
+        return _guess_axes(ds[data_positions], train_pulse_ids, unstack=unstack)
```
Comment (takluyver) [inline extra_data/components.py:617]: unstack isn't defined here

**Comment by tmichela** [comment]
`pulse_dim` or `add_pulse_dim` seems clearer to me as it adds a new dimension to the array with the pulse resolved data?

**Comment by takluyver** [comment]
Unless you think of the existing first dimension as pulses, and this is adding a trains dimension.

Naming things is hard!

**Comment by tmichela** [comment]
good point! so `add_train_dim` ? :D

**Comment by daviddoji** [comment]
> good point! so `add_train_dim` ? :D

I'm happy with both.

File: extra_data/tests/test_components.py [hunk 3]
```diff
@@ -104,6 +104,14 @@ def test_get_array_pulse_indexes_reduced_data(mock_reduced_spb_proc_run):
     assert np.isin(arr.coords['pulse'], [1, 7, 15, 23]).all()
 
 
+def test_get_array_as_xarray(mock_fxe_raw_run):
```
Comment (takluyver) [inline extra_data/tests/test_components.py:107]: Naming again - it returns an xarray object either way, so it's not quite right to call this test `_as_xarray`.

File: extra_data/components.py [hunk 4]
```diff
@@ -328,6 +331,8 @@ def get_array(self, key, pulses=by_index[:]):
           Select the pulses to include from each train. by_id selects by pulse
           ID, by_index by index within the data being read. The default includes
           all pulses. Only used for per-train data.
+        unstack_pulses: bool
+          Wheter to separate train and pulse dimensions.
```
Comment (takluyver) [inline extra_data/components.py:335]:
```suggestion
          Whether to separate train and pulse dimensions.
```

File: extra_data/tests/test_components.py [hunk 5]
```diff
@@ -18,6 +18,9 @@ def test_get_array(mock_fxe_raw_run):
     assert arr.shape == (16, 3, 128, 256, 256)
     assert arr.dims == ('module', 'train', 'pulse', 'slow_scan', 'fast_scan')
 
+    arr = det.get_array('image.data', pulses=by_index[0], unstack_pulses=False)
+    assert arr.shape == (16, 384, 256, 256)
```
Comment (takluyver) [inline extra_data/tests/test_components.py:22]: Check the expected dimension names here as well?

**Comment by takluyver** [comment]
LGTM

**Pull request merged**